pkg/eventqueue/eventqueue.go

```diff
@@ -0,0 +1,207 @@
+	// Enqueued onto the Event queue, or if the Event was Enqueued onto an
+	// EventQueue, and the EventQueue on which the Event was scheduled was
+	// closed.
+	Cancelled chan struct{}
+}
+
+// NewEvent returns an Event with all fields initialized.
+func NewEvent(meta interface{}) *Event {
+	return &Event{
+		Metadata:     meta,
+		EventResults: make(chan interface{}, 1),
+		Cancelled:    make(chan struct{}),
+	}
+}
+
+// WasCancelled returns whether the Cancelled channel for the given Event has
+// been closed or not.
+func (q *Event) WasCancelled() bool {
+	select {
+	case <-q.Cancelled:
+		return true
+	default:
+		return false
+	}
+}
+
+// Enqueue pushes the given event onto the EventQueue. If the queue has been
+// stopped, the Event will not be enqueued, and its cancel channel will be
+// closed, indicating that the Event was not ran. This function may block if
+// there is an event being processed by the queue.
+func (q *EventQueue) Enqueue(ev *Event) {
+	select {
+	case <-q.close:
+		close(ev.Cancelled)
+	default:
+		// The events channel may be closed even if an event has been pushed
+		// onto the events channel, as events are consumed off of the events
+		// channel asynchronously! If the EventQueue is closed before this
+		// event is processed, then it will be cancelled.
+		q.events <- ev
+	}
+}
+
+// Run consumes events that have been queued for this EventQueue. It
+// is presumed that the eventQueue is a buffered channel with a length of one
+// (i.e., only one event can be processed at a time). All business logic for
+// handling queued events is contained within this function. If the Metadata
+// within an Event does not implement the EventHandler interface, the function
+// will log an error, and the Event will not be processed. If the event queue is
+// closed, then all events which were queued up, but not processed, are
+// cancelled. Any event which is currently being processed will not be
+// cancelled. It is assumed that the caller handles both cases for an Event
+// (cancel, or result of event) gracefully.
+func (q *EventQueue) Run() {
+	q.eventQueueOnce.Do(func() {
+		for {
+			select {
+			// Receive next event. No other goroutine or process should consume
+			// events off of this channel!
+			case e := <-q.events:
+				{
+					switch t := e.Metadata.(type) {
+					case EventHandler:
+						ev := e.Metadata.(EventHandler)
+						evRes := ev.Handle()
+						e.EventResults <- evRes
+					default:
+						log.Errorf("unsupported function type provided to event queue: %T", t)
+						// TODO - cancel the event here?
+					}
+
+					// Ensures that no more results can be sent as the event has
+					// already been processed.
+					close(e.EventResults)
+				}
+			// Cancel all events that were not yet consumed.
+			case <-q.close:
+				{
+					// Drain queue of all events.
+					for {
+						select {
+						case drainEvent := <-q.events:
+							close(drainEvent.Cancelled)
+							log.Info("event cancelled!")
```

Comment:
Is this message useful without context? Maybe a leftover?

Reply:
Will remove.